src/number_theory/arithmetic_function.lean

```diff
@@ -221,15 +232,294 @@ instance : semiring (arithmetic_function α) :=
+      { rw [← hcd.1.1, ← hcd.2.1] at cop,
+        rw [← hcd.1.1, h.2, mul_comm, nat.gcd_mul_left,
+            cop.coprime_mul_right.coprime_mul_left_right.gcd_eq_one, mul_one] } },
+    { transitivity nat.gcd (b1 * b2) (a1 * b1),
+      {rw [mul_comm, nat.gcd_mul_right,
+           cop.coprime_mul_right.coprime_mul_left_right.symm.gcd_eq_one, one_mul]},
+      { rw [← hcd.1.1, ← hcd.2.1] at cop,
+        rw [← hcd.2.1, h.1, mul_comm c1 d1, nat.gcd_mul_left,
+            cop.coprime_mul_right.coprime_mul_left_right.symm.gcd_eq_one, mul_one] } },
+    { transitivity nat.gcd (b1 * b2) (a2 * b2),
+      {rw [nat.gcd_mul_right,
+           cop.coprime_mul_left.coprime_mul_right_right.symm.gcd_eq_one, one_mul]},
+      { rw [← hcd.1.1, ← hcd.2.1] at cop,
+        rw [← hcd.2.1, h.2, nat.gcd_mul_right,
+            cop.coprime_mul_left.coprime_mul_right_right.symm.gcd_eq_one, one_mul] } } },
+  { rintros ⟨b1, b2⟩ h,
+    simp only [mem_divisors_antidiagonal, ne.def, mem_product] at h,
+    use ((b1.gcd m, b2.gcd m), (b1.gcd n, b2.gcd n)),
+    simp only [exists_prop, prod.mk.inj_iff, ne.def, mem_product, mem_divisors_antidiagonal],
+    rw [← cop.gcd_mul _, ← cop.gcd_mul _, ← h.1, nat.gcd_mul_gcd_of_coprime_of_mul_eq_mul cop h.1,
+        nat.gcd_mul_gcd_of_coprime_of_mul_eq_mul cop.symm _],
+    { rw [nat.mul_eq_zero, decidable.not_or_iff_and_not] at h, simp [h.2.1, h.2.2] },
+    rw [mul_comm n m, h.1] }
+end⟩
+
+end is_multiplicative
+
+section special_functions
+
+/-- The identity on `ℕ` as an `arithmetic_function`.  -/
+def id : arithmetic_function ℕ := ⟨id, rfl⟩
+
+@[simp]
+lemma id_apply {x : ℕ} : id x = x := rfl
+
+/-- `pow k n = n ^ k`, except `pow 0 0 = 0`. -/
+def pow (k : ℕ) : arithmetic_function ℕ := id.ppow k
+
+@[simp]
+lemma pow_apply {k n : ℕ} : pow k n = if (k = 0 ∧ n = 0) then 0 else n ^ k :=
+begin
+  cases k,
+  { simp [pow] },
+  simp [pow, (ne_of_lt (nat.succ_pos k)).symm],
+end
+
+/-- `σ k n` is the sum of the `k`th powers of the divisors of `n` -/
+def sigma (k : ℕ) : arithmetic_function ℕ :=
+⟨λ n, ∑ d in divisors n, d ^ k, by simp⟩
+
+localized "notation `σ` := sigma" in arithmetic_function
+
+@[simp]
+lemma sigma_apply {k n : ℕ} : σ k n = ∑ d in divisors n, d ^ k := rfl
+
+lemma zeta_mul_pow_eq_sigma {k : ℕ} : ζ * pow k = σ k :=
+begin
+  ext,
+  rw [sigma, zeta_mul_apply],
+  apply sum_congr rfl,
+  intros x hx,
+  rw [pow_apply, if_neg (not_and_of_not_right _ _)],
+  contrapose! hx,
+  simp [hx],
+end
+
+lemma is_multiplicative_zeta [semiring R] : is_multiplicative (ζ : arithmetic_function R) :=
+⟨by simp, λ m n cop, begin
+  cases m, {simp},
+  cases n, {simp},
+  rw [zeta_apply_ne (mul_ne_zero _ _), zeta_apply_ne,
+      zeta_apply_ne, mul_one],
+  repeat { apply nat.succ_ne_zero },
+end⟩
+
+lemma is_multiplicative_pow {k : ℕ} :
+  is_multiplicative (pow k) :=
+⟨by cases k; simp, λ m n cop, by cases m; cases k; simp [nat.succ_ne_zero, mul_pow, zero_pow]⟩
+
+lemma is_multiplicative_sigma [semiring R] {k : ℕ} :
```

Comment:
I'm afraid that this lemma doesn't prove that the uncasted sigma is multiplicative. Because `coe : nat -> nat` is not defeq to the identity.
So I would suggest proving this for `nat`, and when you need to apply it for more general `R`, just using `is_multiplicative_sigma.nat_cast` at that point.